samples/powerplay/src/main/cpp/PowerPlayMultiPlayer.cpp

```diff
@@ -95,22 +94,70 @@ void PowerPlayMultiPlayer::triggerUp(int32_t index) {
 }
 
 void PowerPlayMultiPlayer::triggerDown(int32_t index, oboe::PerformanceMode performanceMode) {
-    auto performanceModeChanged = performanceMode != mLastPerformanceMode;
-    if (index >= 0 && index < mNumSampleBuffers) {
-        mSampleSources[index]->setPlayMode(performanceModeChanged);
+    // Validate index is not out of bounds
+    if (index < 0 || index >= mSampleSources.size()) {
+        __android_log_print(ANDROID_LOG_ERROR, TAG, "triggerDown: Invalid index %d", index);
+        return;
+    }
+
+    // Validate the audio stream is not null.
+    if (mAudioStream == nullptr) {
+        __android_log_print(ANDROID_LOG_ERROR, TAG,
+                            "triggerDown: mAudioStream is null after attempting to open.");
+        return;
     }
 
-    if (performanceModeChanged) {
+    // If the performance mode has changed, we need to reopen the stream.
+    if (performanceMode != mLastPerformanceMode) {
         teardownAudioStream();
-        if (!openStream(performanceMode)) {
+
+        // Attempt here to reopen the stream with the new performance mode.
+        const auto result = openStream(performanceMode);
+        if (!result) {
+            // Something went wrong and the stream could not be reopened.
             __android_log_print(ANDROID_LOG_ERROR,
                                 TAG,
                                 "Failed to reopen stream with new performance mode");
             return;
         }
     }
-    if (mAudioStream) {
-        startStream();
+
+
+    const auto currentPerformanceMode = mAudioStream->getPerformanceMode();
+    const auto currentlyPlayingIndex = getCurrentlyPlayingIndex();
+
+    // Assure all other loaded samples are stopped and the play head is reset to zero, avoiding the
+    // currently playing index. Only allow the playback head to reset when the song has changed.
+    for (size_t i = 0; i < mSampleSources.size(); ++i) {
+        if (i != index) mSampleSources[i]->setStopMode(false);
+        else mSampleSources[i]->setPlayMode(currentlyPlayingIndex == i);
```

Comment:
Shouldn't this always play from the first sample when switching song?

Reply:
It does. This is part of the confusing part and using DrumThumper as the base for this, it's slightly less clear whats happening.

the `setPlayMode` method looks like this: 

```
    void setPlayMode(bool fromPause = true) {
        mIsPlaying = true;
        if (fromPause) {
            mCurSampleIndex = 0;
        }
    }
```

This means in order for the playhead to be reset to zero, from pause needs to be true, but that doesn't make much sense, which also lead me to confusion. Now I'd change it to make more sense, but it'll affect DrumThumper as well and I'd like to avoid that.

I can actually remove this entirely and just put false as well and since all samples are reset anyways, it will work the same.

Reply:
What we do what though is if the user just changes the performance mode, we don't want them to lost the playhead position, thus why the `currentlyPlayingIndex == i` is there.